scalding-core/src/main/scala/com/twitter/scalding/Execution.scala

```diff
@@ -380,70 +387,78 @@ object Execution {
       toWriteCache.getOrPromise((cfg, write))
 
     def getOrElseInsertWithFeedback[T](cfg: Config, ex: Execution[T],
-      res: => Future[(T, Counters)]): (Boolean, Future[(T, Counters)]) =
+      res: => (Future[(T, Counters)], CancellationHandler)): (Boolean, (Future[(T, Counters)], CancellationHandler)) =
       // This cast is safe because we always insert with match T types
       cache.getOrElseUpdateIsNew((cfg, ex), res)
-        .asInstanceOf[(Boolean, Future[(T, Counters)])]
+        .asInstanceOf[(Boolean, (Future[(T, Counters)], CancellationHandler))]
 
     def getOrElseInsert[T](cfg: Config, ex: Execution[T],
-      res: => Future[(T, Counters)]): Future[(T, Counters)] =
+      res: => (Future[(T, Counters)], CancellationHandler)): (Future[(T, Counters)], CancellationHandler) =
       getOrElseInsertWithFeedback(cfg, ex, res)._2
   }
 
 
   private[scalding] final case class FutureConst[T](get: ConcurrentExecutionContext => Future[T]) extends Execution[T] {
     protected def runStats(conf: Config, mode: Mode, cache: EvalCache)(implicit cec: ConcurrentExecutionContext) =
-      Trampoline(cache.getOrElseInsert(conf, this,
-        for {
+      Trampoline {
+        lazy val fut = for {
           futt <- toFuture(Try(get(cec)))
           t <- futt
-        } yield (t, Map.empty)))
+        } yield (t, Map.empty[Long, ExecutionCounters])
+
+        cache.getOrElseInsert(conf, this, (fut, CancellationHandler.empty))
+      }
 
     // Note that unit is not optimized away, since Futures are often used with side-effects, so,
     // we ensure that get is always called in contrast to Mapped, which assumes that fn is pure.
   }
   private[scalding] final case class FlatMapped[S, T](prev: Execution[S], fn: S => Execution[T]) extends Execution[T] {
     protected def runStats(conf: Config, mode: Mode, cache: EvalCache)(implicit cec: ConcurrentExecutionContext) =
-      Trampoline.call(prev.runStats(conf, mode, cache)).map { fut1 =>
-        cache.getOrElseInsert(conf, this,
-          for {
-            (s, st1) <- fut1
-            next0 = fn(s)
-            // next0 has not been optimized yet, we need to try
-            next = optimize(conf, next0)
-            (t, st2) <- Trampoline.call(next.runStats(conf, mode, cache)).get
-          } yield (t, st1 ++ st2))
+      Trampoline.call(prev.runStats(conf, mode, cache)).map { case (fut1, cancelHandler1) =>
+        lazy val uncachedFutCancel = for {
+          (s, st1) <- fut1
+          next0 = fn(s)
+          // next0 has not been optimized yet, we need to try
+          next = optimize(conf, next0)
+          (fut, cancelHandler2) = Trampoline.call(next.runStats(conf, mode, cache)).get
+          (t, st2) <- fut
+        } yield ((t, st1 ++ st2), cancelHandler2)
+
+        val futCancel = cache.getOrElseInsert(conf, this,
+          (uncachedFutCancel.map(_._1), CancellationHandler.empty.compose(uncachedFutCancel.map(_._2))))
```

Comment:
again we are using empty.compose. i think it is better to have a method to lift a handle from a future.